### PR TITLE
base: Add warning when we're at risk of running out of memory

### PR DESCRIPTION
**Pull request by bjackman**
Pandas doesn't seem to properly handle running out of memory, and segfaults; this gives you no useful error message.

This error probably isn't too rare, because the default Lisa config only provisions 1Gb of memory in the Vagrantfile. That means a 500k line ftrace file can cause a segfault.

I've added a hacky warning in base.py. Do you think this is appropriate? If not, we could just put this in the documentation and hope people spot the connection, and perhaps something like this hacky warning could go in Lisa instead?

### CODE REVIEW

File: trappy/base.py [hunk 1]
```diff
@@ -190,6 +200,16 @@ def create_dataframe(self):
                 expl_val = trace_parser_explode_array(val, trace_arr_lengths)
                 self.data_array[idx] = expl_val
 
+        bytes_free = _get_free_memory_bytes()
+        if bytes_free is not None:
+            # Pandas has been observed to segfault here when running out of
+            # memory. Let's make a very rough guess at whether that's likely and
+            # print a warning if we think it is.
+            dataframe_size_guess = 20 * 1024 * len(self.data_array)
```
Comment (JaviMerino) [inline trappy/base.py]: Where does the 20 * 1024 come from?  Every row in the data array uses 20Kb?  Can you add that to the comment?

File: trappy/base.py [hunk 2]
```diff
@@ -190,6 +200,16 @@ def create_dataframe(self):
...
+            if (dataframe_size_guess > 0.8 * bytes_free):
+                logging.warning("TRAPpy: Parsing  large trace may cause errors")
```
Comment (JaviMerino) [inline trappy/base.py]: `warnings.warn`

**Comment by JaviMerino** [comment]
500K is not a huge trace.  I guess that any time that I have parsed big traces I have done it from outside of lisa/vagrant.  I am not against merging this, but we should consider adding more memory to vagrant if it is going to complain so easily.

File: trappy/base.py (hunk 1 above)
Comment (bjackman) [inline trappy/base.py]: Yeah, I actually got this by just trying different trace sizes and looking at the overall memory of the process. On reflection I think this is actually more hacky than I can bear so I'm going to do some more thinking, there has to be a better way.

**Comment by bjackman** [comment]
Yep, I suggested we add more memory to the Lisa Vagrantfile, @derkling was more in favour of documenting the issue so that people can add more memory themselves as needed. If I can find a reliable way of detecting this cause of failure (so that users see "you need more memory" instead of just "Segmentation fault (core dumped)", I guess that would be fine.

**Comment by bjackman** [comment]
Pushed a new version, *slightly* less hacky :)

File: trappy/base.py [hunk 3]
```diff
@@ -15,8 +15,23 @@
 
 """Base class to parse trace.dat dumps"""
 
+import logging
```
Comment (JaviMerino) [inline trappy/base.py]: You can drop this now.

**Comment by bjackman** [comment]
Removed the logging check. 

I tested it by constructing an FTrace object with a particular trace file on VMs with varying memory sizes. As I mentioned in the commit message I haven't tested on systems without /proc/meminfo but I did try changing the `"/proc/meminfo"` string in `_get_free_memory_kb` to garbage, and nothing untoward happened.

I also wasn't able to trigger a case where the warning gets printed but the memory error doesn't occur; the closest I could get is using a 67Mb trace file on a system with 1118164 bytes memory; the warning gets printed, Pandas doesn't segfault, but instead we seem to get OOM killed.

**Comment by JaviMerino** [comment]
Thanks! :heart:

**Pull request merged**